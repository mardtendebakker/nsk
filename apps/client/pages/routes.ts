export const DASHBOARD = '/';
export const SIGN_IN = '/sign-in';
