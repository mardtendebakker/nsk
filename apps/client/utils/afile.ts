import { AFile } from './axios/models/aFile';

export function buildAFileLink(afile: AFile): string {
  return `https://${afile.unique_server_filename}.s3.amazonaws.com/${afile.discr}/${afile.original_client_filename}`;
}
