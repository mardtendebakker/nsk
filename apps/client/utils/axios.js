import axios from 'axios';

export { AxiosPromise, AxiosResponse, AxiosError } from 'axios';

const baseURL = process.env.NODE_ENV === 'development' ? 'http://localhost:3333' : 'production_url';

export const client = axios.create({
  baseURL,
  timeout: 120000,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
});

export const { CancelToken } = axios;

export const SUPLIERS_PATH = '';

export default client;
