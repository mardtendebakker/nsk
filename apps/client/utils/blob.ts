export function openBlob(blob: Blob): void {
  const a = document.createElement('a');
  document.body.appendChild(a);
  a.style.cssText = 'display: none';

  const url = window.URL.createObjectURL(blob);

  a.href = url;
  a.target = '_blank';
  a.click();
  window.URL.revokeObjectURL(url);
}
