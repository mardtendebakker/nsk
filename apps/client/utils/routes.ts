export const DASHBOARD = '/';
export const SIGN_IN = '/sign-in';
export const CUSTOMERS_CONTACTS = '/customers/contacts';
export const CUSTOMERS_CONTACTS_NEW = '/customers/contacts/new';
export const CUSTOMERS_EMAILS_NEW = '/customers/emails/new';
export const CUSTOMERS_EMAILS = '/customers/emails';
export const SUPPLIERS = '/suppliers/:id';
export const ACCOUNT_VERIFICATION = '/account-verification';
export const SETTINGS = '/settings';
export const ORDERS_PURCHASES = '/orders/purchases';
export const ORDERS_SALES = '/orders/sales';
export const STOCKS_PRODUCTS = '/stock/products';
export const STOCKS_REPAIR_SERVICES = '/stock/repair-services';
