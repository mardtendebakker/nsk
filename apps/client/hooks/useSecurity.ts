import { useEffect, useState } from 'react';
import { State } from '../stores/security/types';
import securityStore, {
  SEND_VERIFICATION_CODE_REQUEST,
  SEND_VERIFICATION_CODE_REQUEST_FAILED,
  SEND_VERIFICATION_CODE_REQUEST_SUCCEEDED,
  CONFIRM_ACCOUNT_REQUEST,
  CONFIRM_ACCOUNT_REQUEST_FAILED,
  CONFIRM_ACCOUNT_REQUEST_SUCCEEDED,
  EVENTS,
  SIGN_IN_REQUEST,
  SIGN_IN_REQUEST_FAILED,
  SIGN_OUT,
  SIGN_UP_REQUEST,
  SIGN_UP_REQUEST_FAILED,
  SIGN_UP_REQUEST_SUCCEEDED,
  SIGN_IN_REQUEST_SUCCEEDED,
} from '../stores/security';
import axiosClient, {
  SIGN_IN_PATH, SIGN_UP_PATH, USER_INFO_PATH, CONFIRM_ACCOUNT_PATH,
} from '../utils/axios';
import useAxios from './useAxios';
import useTranslation from './useTranslation';
import { SEND_VERIFICATION_CODE_PATH } from '../utils/axios/paths';
import buildUserFromResponse from '../utils/axios/buildUserFromResponse';

const useSecurity = (): {
  state: State,
  signIn: (object: { username: string, password: string }) => Promise<void>,
  signUp: (object: { username: string, email: string, password: string }) => Promise<void>,
  confirmAccount: (object: { code: string }) => Promise<void>,
  sendVerificationCode: () => Promise<void>,
  signOut: () => void,
  refreshUserInfo: ()=> Promise<void>,
} => {
  const [state, setState] = useState<State>(securityStore.state);
  const { call } = useAxios('post', SIGN_IN_PATH, { withProgressBar: true });
  const { trans } = useTranslation();
  const { call: signUpCall } = useAxios(
    'post',
    SIGN_UP_PATH,
    { withProgressBar: true, customSuccessMessage: trans('signUpSuccessMessage') },
  );
  const { call: confirmAccountCall } = useAxios(
    'post',
    CONFIRM_ACCOUNT_PATH,
    { withProgressBar: true, customSuccessMessage: trans('confirmAccountSuccessMessage') },
  );
  const { call: sendVerificationCodeCall } = useAxios(
    'post',
    SEND_VERIFICATION_CODE_PATH,
    { withProgressBar: true, showSuccessMessage: true },
  );

  function cb() {
    setState(securityStore.state);
  }

  useEffect(() => {
    EVENTS.forEach((event) => {
      securityStore.on(event, cb);
    });

    return () => {
      EVENTS.forEach((event) => {
        securityStore.removeListener(event, cb);
      });
    };
  }, []);

  return {
    state,
    signOut: () : void => {
      securityStore.emit(SIGN_OUT);
    },
    signIn: async (
      { username, password }: { username: string, password: string },
    ) : Promise<void> => {
      securityStore.emit(SIGN_IN_REQUEST);
      try {
        const response = await call({ body: { username, password } });
        if (response) {
          securityStore.emit(SIGN_IN_REQUEST_SUCCEEDED, buildUserFromResponse(response));
        }
      } catch (e) {
        securityStore.emit(SIGN_IN_REQUEST_FAILED);
        throw e;
      }
    },
    signUp: async (
      { username, email, password }:
      { username: string, email: string, password: string },
    ) => {
      securityStore.emit(SIGN_UP_REQUEST);
      try {
        await signUpCall({ body: { username, email, password } });
        securityStore.emit(SIGN_UP_REQUEST_SUCCEEDED, { username, email, emailVerified: false });
      } catch (e) {
        securityStore.emit(SIGN_UP_REQUEST_FAILED);
        throw e;
      }
    },
    confirmAccount: async ({ code }: { code: string }) => {
      securityStore.emit(CONFIRM_ACCOUNT_REQUEST);
      try {
        await confirmAccountCall({ body: { code, username: state.user.username } });
        securityStore.emit(CONFIRM_ACCOUNT_REQUEST_SUCCEEDED);
      } catch (e) {
        securityStore.emit(CONFIRM_ACCOUNT_REQUEST_FAILED);
        throw e;
      }
    },
    refreshUserInfo: async (): Promise<void> => {
      const response = await axiosClient.get(USER_INFO_PATH);
      securityStore.emit(
        SIGN_IN_REQUEST_SUCCEEDED,
        {
          ...securityStore.state.user,
          username: response.data.username,
          email: response.data.email,
        },
      );
    },
    sendVerificationCode: async (): Promise<void> => {
      securityStore.emit(SEND_VERIFICATION_CODE_REQUEST);
      try {
        await sendVerificationCodeCall({ body: { username: state.user.username } });
        securityStore.emit(SEND_VERIFICATION_CODE_REQUEST_SUCCEEDED);
      } catch (e) {
        securityStore.emit(SEND_VERIFICATION_CODE_REQUEST_FAILED);
        throw e;
      }
    },
  };
};

export default useSecurity;
