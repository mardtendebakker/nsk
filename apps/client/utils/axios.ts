import axios from 'axios';

export { AxiosError } from 'axios';
export type { AxiosPromise, AxiosResponse } from 'axios';

const baseURL = process.env.NODE_ENV === 'development' ? 'http://localhost:3333/api' : process.env.NX_AXIOS_BASE_URL;

export const client = axios.create({
  baseURL,
  timeout: 120000,
  headers: {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  },
});

export const { CancelToken } = axios;

export const CUSTOMERS_PATH = '/customers/:id';
export const SUPPLIERS_PATH = '/suppliers/:id';

export default client;

export interface Customer {
  id: number,
  name?: string,
  representative?: string,
  kvk_nr?: string,
  email?: string,
  phone?: string,
  phone2?: string,
  street?: string,
  street_extra?: string,
  city?: string,
  country?: string,
  state?: string,
  zip?: string,
  street2?: string,
  street_extra2?: string,
  city2?: string,
  country2?: string,
  state2?: string,
  zip2?: string,
  is_partner?: number,
}

export interface Supplier {
  id: number,
  name?: string,
  representative?: string,
  email?: string,
  phone?: string,
  phone2?: string,
  street?: string,
  street_extra?: string,
  city?: string,
  country?: string,
  state?: string,
  zip?: string,
  street2?: string,
  street_extra2?: string,
  city2?: string,
  country2?: string,
  state2?: string,
  zip2?: string,
  partner?: string,
}
