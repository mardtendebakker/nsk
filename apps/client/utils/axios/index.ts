import axios from 'axios';
import axiosClient from './client';
import refreshToken from './refreshTokenInterceptor';

export {
  CUSTOMERS_PATH,
  SUPPLIERS_PATH,
  SIGN_IN_PATH,
  REFRESH_TOKEN_PATH,
  USER_INFO_PATH,
  SIGN_UP_PATH,
  CONFIRM_ACCOUNT_PATH,
  CHANGE_PASSWORD_PATH,
  SEND_VERIFICATION_CODE_PATH,
  FORGOT_PASSWORD_PATH,
  SALES_ORDERS_PATH,
  PURCHASE_ORDERS_PATH,
  STOCK_PRODUCTS_PATH,
  USERS_PATH,
  STOCK_REPAIR_SERVICES_PATH,
  ORDER_STATUSES_PATH,
  COMPANIES_PATH,
  LOCATIONS_PATH,
  BULK_PRINT_SALES_PATH,
  BULK_PRINT_PURCHASES_PATH,
  ORDERS_ANALYTICS_PATH,
} from './paths';
export { AxiosError } from 'axios';
export type { AxiosPromise, AxiosResponse } from 'axios';

axiosClient.interceptors.response.use((response) => response, refreshToken);

export const { CancelToken } = axios;

export default axiosClient;
