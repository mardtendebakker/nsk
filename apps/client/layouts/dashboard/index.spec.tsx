import { render } from '@testing-library/react';
import useSecurity from '../../hooks/useSecurity';
import { SIGN_IN, ACCOUNT_VERIFICATION } from '../../utils/routes';
import DashboardLayout from './index';

jest.mock('./header', () => () => {});

const mockRouter = {
  push: jest.fn(),
};

jest.mock('next/router', () => ({
  useRouter: () => mockRouter,
}));

const mockUseSecurity = {
  state: {
    user: { get emailVerified() { return true; } },
  },
  refreshUserInfo: jest.fn(() => Promise.resolve()),
};

jest.mock('../../hooks/useSecurity', () => jest.fn(() => mockUseSecurity));

describe('DashboardLayout', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<DashboardLayout><b id="children">Text</b></DashboardLayout>);
    expect(asFragment()).toMatchSnapshot();
  });
  it('should render successfully when email verified', () => {
    const { container } = render(<DashboardLayout><b id="children">Text</b></DashboardLayout>);
    const element = container.querySelector('#children');
    const { refreshUserInfo } = useSecurity();

    expect(refreshUserInfo).toBeCalled();
    expect(element).toBeDefined();
  });
  it('should redirects to account verification when email not verified', () => {
    jest.spyOn(mockUseSecurity.state.user, 'emailVerified', 'get').mockReturnValue(false);

    const { container } = render(<DashboardLayout><b id="children">Text</b></DashboardLayout>);
    const element = container.querySelector('#children');

    expect(element).toBeNull();
    expect(mockRouter.push).toBeCalledWith(ACCOUNT_VERIFICATION);
  });
  it('should redirects to sign in when email not verified', () => {
    useSecurity.mockImplementation(() => ({
      state: { user: null },
      refreshUserInfo: jest.fn(() => Promise.resolve()),
    }));
    const { container } = render(<DashboardLayout><b id="children">Text</b></DashboardLayout>);
    const element = container.querySelector('#children');

    expect(element).toBeNull();
    expect(mockRouter.push).toBeCalledWith(SIGN_IN);
  });
});
