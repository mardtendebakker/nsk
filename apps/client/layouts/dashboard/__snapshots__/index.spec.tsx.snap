// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardLayout matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-5pl8f4"
  >
    <main
      class="MuiBox-root css-1nm38x2"
    >
      <div
        class="MuiContainer-root css-ur2jdm-MuiContainer-root"
      >
        <b
          id="children"
        >
          Text
        </b>
      </div>
    </main>
  </div>
</DocumentFragment>
`;
